torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -721,6 +721,8 @@ def reset_sharded_param(self):
         self._sharded_param_data = local_tensor.view(-1)
         assert isinstance(self.sharded_param, DTensor)  # mypy
         self.sharded_param._local_tensor = local_tensor[: self.sharded_size[0]]
+        if torch.distributed.get_rank() == 0:
+            print(f"{self.sharded_param.device.type=} {self.sharded_param._local_tensor.device.type=}")
```

Comment:
It means DTensor.device is still cuda. This prints `self.sharded_param.device.type='cuda' self.sharded_param.device.type='cuda'`.

Reply:
@awgu curious if you have any suggestions. this is a minimized repro of torchtune bug: https://github.com/pytorch/torchtune/issues/1412

I tried following but it's problematic because optimizer.param_groups refers to a different `model.parameters()`
```
self.sharded_param = nn.Parameter(
  self.to_sharded_dtensor(local_tensor[: self.sharded_size[0]]),
  requires_grad=new_param.requires_grad,
)
self._setattr_on_modules(self.sharded_param)
```

Reply:
`reset_sharded_param` should run in the post-load hook. Why does the optimizer already have a reference to the pre-load parameters? Or, am I misunderstanding?

Reply:
I think I will need to understand the optimizer constraint to propose my thoughts on what a solution should be.

Reply:
figured out: the problamatic part is calling `reset_sharded_param` from `lazy_init`. I should avoid creating `nn.parameters` if they are the same https://github.com/pytorch/pytorch/pull/135156

trunk is healthy because we use `._local_tensor=`. It only matters if I create a new `nn.Pararmeter` https://github.com/pytorch/pytorch/pull/135156

closing this PR so we can review the new PR